src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml

```diff
@@ -188,6 +189,27 @@
 		</property>
 	</bean>
 
+	<bean id="awsSecretValidationJob"
+		parent="unscheduledJobDescriptor">
+		<property name="steps">
+			<list>
+				<ref bean="awsSecretValidationStep" />
+			</list>
+		</property>
+	</bean>
+
+	<bean id="awsSecretValidationStep" class="org.kuali.kfs.sys.aws.batch.AwsSecretValidationStep" parent="step" >
+		<property name="awsSecretService" ref="awsSecretService" />
+	</bean>
+
+	<bean id="awsSecretService" parent="awsSecretService-parentBean" />
+	<bean id="awsSecretService-parentBean" abstract="true" class="edu.cornell.kfs.sys.service.impl.AwsSecretServiceImpl"
+		scope="prototype"
```

Comment:
Shouldn't the "awsSecretService" bean be declared with prototype scope as well?

Also, I'm trying to figure out how the prototype setup would work out when put into actual non-testing use, as well as what the expectations are for the cache lifecycle. Would each dependent service have its own AwsSecretService instance that is only initialized at startup, and then the service's processing will call the cache management methods on its own AwsSecretService instance as needed? Or would the dependent services be expected to retrieve a new AwsSecretService instance prior to starting their next processing run? Or would the dependent services need to be declared as prototypes as well, and then the caller would have to retrieve a new instance of that service?

Example case: ConcurReportsServiceImpl is dependent on ConcurAccessTokenServiceImpl to retrieve the current access token prior to making its calls to Concur. If we update this process to use AWS secrets, ConcurAccessTokenServiceImpl would have its own AwsSecretService instance. ConcurReportsServiceImpl would then be affected by the secrets caching of ConcurAccessTokenServiceImpl, unless we update ConcurReportsServiceImpl to have its own AwsSecretService to retrieve the access token independently.

Reply:
I'll add the prototype scope to the awsSecretService bean.

I wanted to configure this such that each time Spring delivers the awsSecretService, it will be a new instance, thus creating a separate cache.  This would also provide the flexibility of each service needing to use awsSecretService, could handle the cache to its own needs.

In the case of Concur, my expectation is the following.  ConcurReportsServiceImpl and ConcurAccessTokenServiceImpl would both have awsSecretService injected with Spring.  First we go through ConcurAccessTokenServiceImpl to update the tokens.  Then we would let ConcurReportsServiceImpl perform its functions.  Both services should clear cache in awsSecretService before and after they complete processing.  Doing so would make sure both have the latest secret tokens.  Do you think there is a better approach?

Reply:
Having each service use its own AwsSecretService instance sounds good in general. With such an approach, though, the dependent services may have to add a new method for invoking the cache-clearing method on their internal AwsSecretService instance.

For example, ConcurEventNotificationService and/or ConcurEventNotificationProcessingService will invoke methods on ConcurReportsService that make use of the access token. Once we update our Concur process to use AWS secrets, a new ConcurReportsService method for clearing the AWS cache might be needed. That way, the Concur batch step class (or one of the notification services) can clear the cache at the appropriate times.

I can't think of a better overall approach than what you had suggested. The GlobalVariables-like approach from my earlier PR ended up being too complex and not as clear when I presented it to the team.